pmlc/dialect/pxa/analysis/affine_constraints.cc

```diff
@@ -0,0 +1,210 @@
+// Copyright 2020, Intel Corporation
+#include "pmlc/dialect/pxa/analysis/affine_constraints.h"
+
+#include <algorithm>
+#include <functional>
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+#include "pmlc/dialect/pxa/ir/interfaces.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::dialect::pxa {
+
+mlir::LogicalResult
+addAffineParallelIVDomain(mlir::AffineParallelOp parallelOp, unsigned idx,
+                          mlir::FlatAffineConstraints &constraints) {
+  unsigned pos;
+  if (!constraints.findId(parallelOp.getIVs()[idx], &pos)) {
+    assert(false && "Value not found");
+    return mlir::failure();
+  }
+
+  mlir::AffineValueMap lowerMap = parallelOp.getLowerBoundsValueMap();
+  mlir::AffineValueMap upperMap = parallelOp.getUpperBoundsValueMap();
+  if (auto constLower =
+          lowerMap.getResult(idx).cast<mlir::AffineConstantExpr>())
+    constraints.addConstantLowerBound(pos, constLower.getValue());
+  if (auto constUpper =
+          upperMap.getResult(idx).cast<mlir::AffineConstantExpr>())
+    constraints.addConstantUpperBound(pos, constUpper.getValue() - 1);
+  return mlir::success();
+}
+
+mlir::Optional<int64_t>
+getLowerBound(mlir::AffineExpr expr, mlir::FlatAffineConstraints &constraints) {
+  return getLowerUpperBounds(expr, constraints).first;
+}
+
+mlir::Optional<int64_t>
+getUpperBound(mlir::AffineExpr expr, mlir::FlatAffineConstraints &constraints) {
+  return getLowerUpperBounds(expr, constraints).second;
+}
+
+namespace {
+
+template <typename T, typename BinOp>
+auto applyOptional(const mlir::Optional<T> &a, const mlir::Optional<T> &b,
+                   BinOp &&op)
+    -> mlir::Optional<decltype(op(a.getValue(), b.getValue()))> {
+  if (a.hasValue() && b.hasValue())
+    return op(a.getValue(), b.getValue());
+  return llvm::None;
+}
+
+template <typename T>
+struct minimum {
+  T operator()(const T &lhs, const T &rhs) { return std::min(lhs, rhs); }
+};
+
+template <typename T>
+struct maximum {
+  T operator()(const T &lhs, const T &rhs) { return std::max(lhs, rhs); }
+};
+
+} // namespace
+
+std::pair<mlir::Optional<int64_t>, mlir::Optional<int64_t>>
+getLowerUpperBounds(mlir::AffineExpr expr,
```

Comment:
This code is heavily overlapping in purpose with StrideRange in pxa/analysis, but this code is better I think.  Specifically, it more properly supports semi-affine operations like mod.  It is probably worth considers how to dedup on some later commit.